src/raftstore/store/store.rs

```diff
@@ -1540,6 +1704,80 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         }
     }
 
+    fn get_slibing_peer(
+        &self,
+        region: &metapb::Region,
+        direction: MergeDirection,
+    ) -> Option<&Peer> {
+        match direction {
+            MergeDirection::Forward => {
+                let region_start_key = enc_end_key(region);
+                self.region_ranges
+                    .range((Excluded(region_start_key), Unbounded::<Key>))
+                    .next()
+                    .and_then(|(_, id)| {
+                        let peer = match self.region_peers.get(id) {
+                            None => return None,
+                            Some(p) => p,
+                        };
+                        if enc_start_key(peer.region()) == enc_end_key(region) {
+                            return Some(peer);
+                        }
+                        None
+                    })
+            }
+            MergeDirection::Backward => {
+                let slibing_region_key = enc_start_key(region);
+                self.region_ranges
+                    .get(&slibing_region_key)
+                    .and_then(|region_id| self.region_peers.get(region_id))
+            }
+        }
+    }
+
+    fn check_merge_proposal(&self, msg: &mut RaftCmdRequest) -> Result<()> {
+        if !msg.get_admin_request().has_pre_merge() && !msg.get_admin_request().has_merge() {
+            return Ok(());
```

Comment:
Why not return an error? It is't a merge proposal after all.

Reply:
This method is called for all requests, and skip all non-merge requests.